docs/proposals/pod-resource-management.md

```diff
@@ -0,0 +1,394 @@
+
+# Pod level resource management in Kubelet
+
+**Author**: Buddha Prakash (@dubstack), Vishnu Kannan (@vishh)
+
+**Last Updated**: 06/23/2016
+
+**Status**: Draft Proposal (WIP)
+
+This document proposes a design for introducing pod level resource accounting to Kubernetes, and outlines the implementation and rollout plan.
+
+<!-- BEGIN MUNGE: GENERATED_TOC -->
+
+- [Pod level resource management in Kubelet](#pod-level-resource-management-in-kubelet)
+  - [Introduction](#introduction)
+  - [Non Goals](#non-goals)
+  - [Motivations](#motivations)
+  - [Design](#design)
+    - [Proposed cgroup hierarchy:](#proposed-cgroup-hierarchy)
+      - [Quality of Service classes](#quality-of-service-classes)
+      - [Guaranteed](#guaranteed)
+      - [Burstable](#burstable)
+      - [Best Effort](#best-effort)
+    - [With Systemd](#with-systemd)
+    - [Hierarchy Outline](#hierarchy-outline)
+  - [Implementation Plan](#implementation-plan)
+      - [Top level Cgroups for QoS tiers](#top-level-cgroups-for-qos-tiers)
+      - [Pod level Cgroup creation and deletion (Docker runtime)](#pod-level-cgroup-creation-and-deletion-docker-runtime)
+      - [Container level cgroups](#container-level-cgroups)
+      - [Rkt runtime](#rkt-runtime)
+      - [Add Pod level metrics to Kubelet's metrics provider](#add-pod-level-metrics-to-kubelets-metrics-provider)
+  - [Rollout Plan](#rollout-plan)
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+## Introduction
+
+As of now [Quality of Service](../../docs/design/resource-qos.md) is not enforced at a pod level. Excepting pod evictions, all the other QoS features are not applicable at the pod level.
+To better support QoS, there is a need to add support for pod level resource accounting in Kubernetes.
+
+We propose to have to have a unified cgroup hierarchy with pod level cgroups for better resource management. We will have a cgroups hierarchy with top level cgroups for the three Quality of Service(QoS) classes Guaranteed, Burstable and BestEffort. Pods (and their containers) belonging to a QoS class will be grouped under these top level QoS cgroups. And all containers in a pod are nested under the pod cgroups.
+
+The proposed cgroup hierarchy would allow for more efficient resource management and lead to improvements in node reliability.
+This would also allow for significant latency optimizations in terms of pod eviction on nodes with the use of pod level resource usage metrics.
+This document provides a basic outline of how we plan to implement and rollout this feature.
+
+
+## Non Goals
+
+- Pod level disk accounting will not be tackled in this proposal.
+- Pod level resource specification in the Kubernetes API will not be tackled in this proposal.
+
+## Motivations
+
+Kubernetes currently supports container level isolation only and lets users specify resource requests/limits on the containers [Compute Resources](../../docs/design/resources.md). Kubelet creates cgroups sandbox (via docker) for all these containers.
```

Comment:
The `kubelet` creates a cgroup sandbox (via it's container runtime) for each container.